### PR TITLE
Check even or odd challenge

### PR DESCRIPTION
**Pull request by duhaluay**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/even_or_odd_py
Comment (hassanql) [inline solutions/even_or_odd_py:1]: The function is_even is well-written and straightforward, but there are several improvements needed to meet the coding standards outlined in the review checklist. First, the function lacks type annotations and defensive assertions to handle invalid inputs. Adding a type check for the num parameter (e.g., if not isinstance(num, int): raise TypeError("Input must be an integer.")) would enhance robustness. Additionally, the print statement at the end of the file should be removed, as function calls should be handled in a test file, not the main function file. The docstring is clear, but adding doctests with examples, including edge cases and invalid inputs, would improve clarity and test coverage and was required (refer to the solutions other team members have uploaded). take this as an example:
"""
    >>> is_even(2)
    True
    >>> is_even(3)
    False
    >>> is_even(0)
    True
    >>> is_even("a")
    Traceback (most recent call last):
    ...
    TypeError: Input must be an integer.
"""

A corresponding test file (test_is_even.py) is also missing, which should include unit tests for standard cases (e.g., is_even(2)), edge cases (e.g., is_even(0)), and invalid cases (e.g., is_even("a")). These tests should follow proper naming conventions and include clear docstrings.

**Review by hassanql — CHANGES_REQUESTED** [review]

File: solutions/is_even.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+"""
+This module contains a function that checks if a number is even or odd.
```
Comment (hassanql) [inline solutions/is_even.py:2]: looks good. Please add a module level docstring like the one you put on test_is_even.py. You can follow this format:
"""
Created: [DD/MM/YY]
Team Number: 28
Team Name: MIT Alpha
Author: [Your Name]
"""

File: solutions/even_or_odd_py
Comment (duhaluay) [reply solutions/even_or_odd_py:1]: DONE

File: solutions/is_even.py (hunk 1 above)
Comment (duhaluay) [reply solutions/is_even.py:2]: Added the meta data

**Review by hassanql — APPROVED** [review]
All good now. Thank you for encoperating the changes.
